### PR TITLE
21187 add merge selection functionality to old sans gui

### LINKED ISSUE #21187: Add merge selection functionality to old sans GUI.

### PR DESCRIPTION
**Pull request by Matthew-Andrew**
This PR adds the functionality whereby the merge region can be specified to the old SANS GUI. This should be merged after the PR which implements this on the backend and in the new GUI #21107. 

**To test:**
- Get the SANS data from http://download.mantidproject.org/. The ISIS sample data has some SANS data in a folder called loqdemo.
- Add this folder to your directories
- Open mantid, go to interfaces->SANS->ISIS SANS 
- Load the maskfile.txt from the folder as a user file.
- Load the batch_mode_reduction.csv as the batch file
- Go to the reduction tab and select merged reduction 
- Tick Use Range Box, put 0.1 and 0.2 in the two boxes to the left.
- Go to main tab and click 1D Reduce
- Plot all three output workspaces
- Confirm that the merged reduction matches LAB below 0.1, HAB above 0.2 and is an average of the two between these two values.

Code review

Fixes #21187 . 

**Release Notes** 
Release notes already added in #21107

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by soininen** [comment]
I tried to follow the instructions. The output workspaces are `first_time_hab`, `first_time_main` and `first_time_merged`, right?  The average of the single Y value at 0.158 1/Å for `first_time_hab` and `first_time_main` is ~0.149, while the corresponding Y in `first_time_merged` is ~0.152. Is the discrepancy expected or am I doing something wrong here?

**Comment by Mantid-Matthew** [comment]
You haven't done anything wrong. I misused the word average in the instructions. The merged data is not actually a simple average of the front and rear detectors. It is actually given by (In the simplest case) (Cf + Cr)/(Nr+Nf). Where C is the sum of counts and N is the sum of norms. The individual reductions are given by there own C/N respectively. I meant the last line of the instructions to be a simple check that the merged reduction lies between the HAB and the LAB reductions apologies.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSStitch.py [hunk 1]
```diff
@@ -335,6 +333,13 @@ def PyExec(self):
         nF = self._crop_to_x_range(nF, min_q, max_q)
         nR = self._crop_to_x_range(nR, min_q, max_q)
 
+        if merge_mask and self.getProperty('ProcessCan').value:
+            cR_can, cF_can, nR_can, nF_can = self._apply_user_mask_ranges(cF_can, cR_can, nR_can, nF_can, merge_min,
+                                                                          merge_max)
+
+        if merge_mask:
+            cR, cF, nR, nF = self._apply_user_mask_ranges(cF, cR, nR, nF, merge_min, merge_max)
```
Comment (soininen) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSStitch.py]: Could the following be a bit clearer:
```python
if merge_mask:
    if self.getProperty('ProcessCan').value:
        cR_can, cF_can, nR_can, nF_can = self._apply_user_mask_ranges(cF_can, cR_can, nR_can, nF_can, merge_min, merge_max)
    else:
        cR, cF, nR, nF = self._apply_user_mask_ranges(cF, cR, nR, nF, merge_min, merge_max)
```

File: scripts/SANS/ISISCommandInterface.py [hunk 2]
```diff
@@ -218,6 +218,22 @@ def SetFrontDetRescaleShift(scale=1.0, shift=0.0, fitScale=False, fitShift=False
...
+def SetMergeQRange(q_max=None, q_min=None):
+    """
+        Stores property about the detector which is used to rescale and shift
+        data in the bank after data have been reduced
+        @param scale: Default to 1.0. Value to multiply data with
+        @param shift: Default to 0.0. Value to add to data
+        @param fitScale: Default is False. Whether or not to try and fit this param
+        @param fitShift: Default is False. Whether or not to try and fit this param
+        @param qMin: When set to None (default) then for fitting use the overlapping q region of front and rear detectors
+        @param qMax: When set to None (default) then for fitting use the overlapping q region of front and rear detectors
```
Comment (soininen) [inline scripts/SANS/ISISCommandInterface.py]: What are these params? I only see `q_max` and `q_min` in the signature above.

Also: I think it is also customary to order the min and max arguments as `def SetMergeQRange(q_min, q_max)`. Is there a specific reason why they are the other way here?

File: scripts/SANS/isis_instrument.py [hunk 3]
```diff
@@ -135,6 +135,23 @@ def isRectangle(self):
         def n_pixels(self):
             return self._n_pixels
 
+    class _MergeRange(object):
+        """
+            Stores property about the detector which is used specify the merge ranges after the data has been reduced.
```
Comment (soininen) [inline scripts/SANS/isis_instrument.py]: used **to** specify?

File: scripts/SANS/isis_instrument.py [hunk 4]
```diff
@@ -135,6 +135,23 @@ def isRectangle(self):
...
+        """
+        def __init__(self, q_max=None, q_min=None):
```
Comment (soininen) [inline scripts/SANS/isis_instrument.py]: Here, as well, would it make more sense to have min before max?

**Review by soininen — CHANGES_REQUESTED** [review]
As far as I can tell, interface works as expected. There were a few minor issues in the code, but overall looks like a nice change.

**Review by soininen — APPROVED** [review]

**Pull request merged**